### PR TITLE
Update policy migration state (AAD 3.4)

### LINKED ISSUE #1730: Update MS.AAD.3.4v1 to account for new tenants that can't migrate
### Prerequisites

- [x] This issue has an informative and human-readable title.

### 💡 Summary

The purpose of the request is to update the REGO to account for new tenants that are unable to migrate and return a NULL value when checking the status.

### Motivation and context

This reduces false positives and accurately captures the status of all tenants regardless of when they were created.

### Implementation notes

1. Add the below code to the `AADConfig.rego` file.

```REGO
AuthenticationPolicyMigrationIsComplete if {
  is_null(AuthenticationPolicyMigrationState)
}
```

### Acceptance criteria

- [ ] AAD rego file updated
- [ ] Sufficient test coverage is added

### PR DESCRIPTION
**Pull request by MichaelHicks-MSFT**
#  #
  

## 🗣 Description ##
Added additional values to account for tenants that are unable to migrate due to having newer tenants. Newer tenants will have null listed within the `ScubaResults_*****.json` file instead of `migrationComplete`.

Closes #1730

## 🧪 Testing ##
Run a scan of a test tenant. If the tenant doesn't have null listed within the results file, edit the file then run Invoke-SCuBACached to verify it passes.

```json
"authentication_method_policy":  {
    "AuthenticationMethodConfigurations@odata.context":  "https://graph.microsoft.com/beta/$metadata#policies/authenticationMethodsPolicy/authenticationMethodConfigurations",
    "PolicyVersion":  "1.5",
    "PolicyMigrationState":  null,
    "DisplayName":  "Authentication Methods Policy",
```

## ✅ Pre-approval checklist ##

- [x] This PR has an informative and human-readable title.
- [x] PR targets the correct parent branch (e.g., main or release-name) for merge.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] Changes are sized such that they do not touch excessive number of files.
- [x] *All* future TODOs are captured in issues, which are referenced in code comments.
- [x] These code changes follow the ScubaGear [content style guide](https://github.com/cisagov/ScubaGear/blob/main/CONTENTSTYLEGUIDE.md).
- [x] Related issues these changes resolve are linked preferably via [closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).
- [x] All relevant type-of-change labels added.
- [x] All relevant project fields are set.
- [x] All relevant repo and/or project documentation updated to reflect these changes.
- [x] Unit tests added/updated to cover PowerShell and Rego changes.
- [x] Functional tests added/updated to cover PowerShell and Rego changes.
- [ ] All relevant functional tests passed.
- [x] All automated checks (e.g., linting, static analysis, unit/smoke tests) passed.

## ✅ Pre-merge checklist ##

- [ ] PR passed smoke test check.
- [ ] Feature branch has been rebased against changes from parent branch, as needed

  Use `Rebase branch` button below or use [this](https://www.digitalocean.com/community/tutorials/how-to-rebase-and-update-a-pull-request) reference to rebase from the command line.
- [ ] Resolved all merge conflicts on branch
- [ ] Notified merge coordinator that PR is ready for merge via comment mention
- [ ] Demonstrate changes to the team for questions and comments. 
    (Note: Only required for issues of size `Medium` or larger)

## ✅ Post-merge checklist ##

- [ ] Feature branch deleted after merge to clean up repository.
- [ ] Verified that all checks pass on parent branch (e.g., main or release-name) after merge.

### CODE REVIEW

**Comment by mitchelbaker-cisa** [comment]
Updated priority to high since the issue improves the assessment for a SHALL policy and resolves a known false positive.

File: PowerShell/ScubaGear/Rego/AADConfig.rego [hunk 1]
```diff
@@ -366,6 +366,7 @@ AuthenticationPolicyMigrationState := PolicyMigrationState if {
 # Returns true if the tenant has completed their authpolicy migration
 default AuthenticationPolicyMigrationIsComplete := false
 AuthenticationPolicyMigrationIsComplete if AuthenticationPolicyMigrationState == "migrationComplete"
+AuthenticationPolicyMigrationIsComplete if is_null(AuthenticationPolicyMigrationState)
```
Comment (mitchelbaker-cisa) [inline PowerShell/ScubaGear/Rego/AADConfig.rego:369]: What is the result we're expecting when AuthenticationPolicyMigrationState == null? If I change the setting to null then AAD 3.4 fails the check. Should AAD 3.4 pass if AuthenticationPolicyMigrationState == null?

I assume if the migration status setting does not exist for new tenants (and therefore their migration is "complete") then they should pass AAD 3.4.

**Review by mitchelbaker-cisa — CHANGES_REQUESTED** [review]
We have functional tests for "preMigration" and "migrationComplete". Can we add tests for "migrationInProgress" and "null" values?

```
- TestDescription: MS.AAD.3.4v1 Non-compliant case - migrationInProgress
        Preconditions:
          - Command: UpdateProviderExport
            Splat:
              updates:
                authentication_method[0].authentication_method_policy.PolicyMigrationState: migrationInProgress
        Postconditions: []
        ExpectedResult: false
- TestDescription: MS.AAD.3.4v1 Non-compliant case - null 
        Preconditions:
          - Command: UpdateProviderExport
            Splat:
              updates:
              authentication_method[0].authentication_method_policy.PolicyMigrationState: null
        Postconditions: []
        ExpectedResult: false
```

File: PowerShell/ScubaGear/Rego/AADConfig.rego (hunk 1 above)
Comment (mitchelbaker-cisa) [reply PowerShell/ScubaGear/Rego/AADConfig.rego:369]: Apologies, I was setting AuthenticationPolicyMigrationState equal to `"null"`, a string, vs `null`, the value. When set to `null` AAD 3.4 passes.

**Comment by mitchelbaker-cisa** [comment]
Looks good from my end, will approve after the two functional tests are added.

File: PowerShell/ScubaGear/baselines/aad.md [hunk 2]
```diff
@@ -254,6 +254,7 @@ The Authentication Methods Manage Migration feature SHALL be set to Migration Co
 <!--Policy: MS.AAD.3.4v1; Criticality: SHALL -->
 - _Rationale:_ To disable the legacy authentication methods screen for the tenant, configure the Manage Migration feature to Migration Complete. The MFA and Self-Service Password Reset (SSPR) authentication methods are both managed from a central admin page, thereby reducing administrative complexity and potential security misconfigurations.
 - _Last modified:_ June 2023
+- _Note:_ Migration is not possible for newer tenants, as they use the newer authentication methods by default.
```
Comment (mitchelbaker-cisa) [inline PowerShell/ScubaGear/baselines/aad.md]: What is the specific date where new tenants begin to use modern authentication methods by default? Based on [Migration between Policies documentation](https://learn.microsoft.com/en-us/entra/identity/authentication/concept-authentication-methods-manage#migration-between-policies) and what's seen in the Azure portal, is the official date September 30th, 2025? Is it prior to this date since we've seen customers already encounter the false positive?

I'd prefer the note to say, "This policy is only applicable for tenants created prior to the <September 30th, 2025/insert date) deprecation date for legacy multifactor authentication and self-service password reset policies.

One other thing, this will require a baseline update and we are past our internal deadline for submitting new baseline updates for Nemo. Can we break out this note (plus any new documentation referenced in resources) to a separate issue?

File: Testing/Functional/Products/TestPlans/aad.testplan.yaml [hunk 3]
```diff
@@ -315,6 +315,22 @@ TestPlan:
                 authentication_method[0].authentication_method_policy.PolicyMigrationState: migrationComplete
         Postconditions: []
         ExpectedResult: true
+      - TestDescription: MS.AAD.3.4v1 Non-compliant case - migrationInProgress
```
Comment (mitchelbaker-cisa) [inline Testing/Functional/Products/TestPlans/aad.testplan.yaml:318]: Functional tests look good, did a test run here: https://github.com/cisagov/ScubaGear/actions/runs/17000441529

File: Testing/Functional/Products/TestPlans/aad.testplan.yaml [hunk 4]
```diff
@@ -315,6 +315,22 @@ TestPlan:
...
+                authentication_method[0].authentication_method_policy.PolicyMigrationState: migrationInProgress
+        Postconditions: []
+        ExpectedResult: false
+      - TestDescription: MS.AAD.3.4v1 Compliant case - null
+        Preconditions:
+          - Command: UpdateProviderExport
+            Splat:
+              updates:
+                authentication_method[0].authentication_method_policy.PolicyMigrationState: null
+              Postconditions: []
```
Comment (mitchelbaker-cisa) [inline Testing/Functional/Products/TestPlans/aad.testplan.yaml]: `Postconditions: []` and `ExpectedResult: []` indenting is a bit too far causing this test to fail, they should have the same indentation as `TestDescription` and `Preconditions:`

Comment (mitchelbaker-cisa) [reply Testing/Functional/Products/TestPlans/aad.testplan.yaml]: @MichaelHicks-MSFT 

Indenting just needs to be updated from this:

image

To this:

image

Comment (MichaelHicks-MSFT) [reply Testing/Functional/Products/TestPlans/aad.testplan.yaml]: Updated in latest commit.

File: PowerShell/ScubaGear/baselines/aad.md (hunk 2 above)
Comment (MichaelHicks-MSFT) [reply PowerShell/ScubaGear/baselines/aad.md]: Removed note, will add in later issue.

**Comment by PaulKuykendall** [comment]
@dagarwal-ecs  awaiting changed from @mitchelbaker-cisa

File: Testing/Functional/Products/TestPlans/aad.testplan.yaml [hunk 5]
```diff
@@ -315,6 +315,24 @@ TestPlan:
...
+              updates:
+                authentication_method[0].authentication_method_policy.PolicyMigrationState: migrationInProgress
+        Postconditions: []
+        ExpectedResult: false
+      - TestDescription: MS.AAD.3.4v1 Compliant case - null
+        Preconditions:
+          - Command: UpdateProviderExport
+            Splat:
+              updates:
+                authentication_method:
```
Comment (mitchelbaker-cisa) [inline Testing/Functional/Products/TestPlans/aad.testplan.yaml:331]: An error is thrown when a property inside of `authentication_method[0]` is set to `null`. Updated to temporarily use `authentication_method` for only this test case which resolves the issue.
